File: server/Model/DTOs/Organization/Department.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace Model.DTOs.Organization
{
    public class DepartmentDto
    {
        public int Id { get; set; }

        public string Name { get; set; } = string.Empty;

        public string Code { get; set; } = string.Empty;

        public string? Description { get; set; }

        public string DepartmentTypeCode { get; set; } = string.Empty;

        public string DepartmentTypeName { get; set; } = string.Empty;

        public int? ParentId { get; set; }

        public string? Address { get; set; }

        public List<DepartmentDto>? Children { get; set; }
    }

    public class DetailDepartmentDto
    {
        public int Id { get; set; }
        [Required]
        [StringLength(100)]
        public required string Name { get; set; }

        [Required]
        [StringLength(50)]
        public required string Code { get; set; }

        [StringLength(500)]
        public string? Description { get; set; }

        [Required]
        [StringLength(30)]
        public required string DepartmentTypeCode { get; set; }

        public int ParentId { get; set; }

        [StringLength(200)]
        public string? Address { get; set; }
        public required string TreePath { get; set; }

    }

    public class DepartmentTreeFilterDto
    {
        public string? DepartmentTypeCode { get; set; }
        public string? SearchTerm { get; set; }
    }
}

File: client/types/sys/Department.ts
import { defaultPaginationFilter, PaginationFilter } from '../base/PaginationFilter';

export type DepartmentDto = {
  id: number;
  name: string;
  code: string;
  description?: string;
  departmentTypeCode: string;
  departmentTypeName: string;
  departmentTypeLevel: number;
  parentId?: number;
  address?: string;
  children?: DepartmentDto[];
};

export type DetailDepartmentDto = {
  id: number;
  name: string;
  code: string;
  description?: string;
  departmentTypeCode: string;
  parentId?: number;
  address?: string;
  treePath: string;
};

export const defaultDetailDepartmentDto: DetailDepartmentDto = {
  id: 0,
  name: '',
  code: '',
  description: '',
  departmentTypeCode: '',
  parentId: undefined,
  address: '',
  treePath: '',
};

export type DepartmentMemberDto = {
  id: number;
  userId: string;
  userName: string;
  fullName?: string;
  avatar?: string;
  email?: string;
  isActive?: boolean;
};

export type DepartmentMemberFilter = PaginationFilter & {
  departmentId: number;
  isShowSubMembers: boolean;
};

export const defaultDepartmentMemberFilter: DepartmentMemberFilter = {
  ...defaultPaginationFilter,
  departmentId: 0,
  isShowSubMembers: true,
};

export type AddDepartmentMemberDto = {
  departmentId: number;
  userIds: string[];
};

export type RemoveDepartmentMemberDto = {
  departmentId: number;
  userIds: string[];
};

export type UserDepartmentCursorFilterDto = {
  searchTerm: string;
  departmentId: number;
  limit: number;
  /// <summary>
  /// cursor by created time
  /// </summary>
  cursor: string | null;
};

export const defaultUserDepartmentCursorFilterDto: UserDepartmentCursorFilterDto = {
  searchTerm: '',
  departmentId: 0,
  limit: 50,
  cursor: null,
};
